### PR TITLE
Acoustic msg devel

### PR DESCRIPTION
**Pull request by CaptKrasno**
Hi all,

I implemented some of the changes to Multibeam Related messages we discussed at our last meeting.

In addition, I have created a new repository for Acoustic Message tools.   To get things started I added a water column viewer. you can find the repo here:  [Acoutic Messages Tools](https://github.com/CaptKrasno/acoustic_msgs_tools)
wc_gui2-2022-02-22_15 21 30

I have converted some logged water column data to the new format.  You can find it here if you want to look at it or try the viewer.  You can download that here:  [https://next.krasno.cloud/s/RnaGswoDbrZWkqQ](https://next.krasno.cloud/s/RnaGswoDbrZWkqQ)  

After this pull request is accepted I will fully support the new message types with my Norbit MBES driver, open source it and make it available for use.

It's worth mentioning how the data is packed in the Watercolumn message as it may not be obvious.

- the actual pixel data is in row-major (i.e beam_no major) order. 
- the pixel data is stored as an array of bytes (uint8).  It should be cast as the type specified in dtype flag.  They are enumerated in the message
- to find the range for each cell:  (sample_number + sample0) * sound_speed / (2.0 * sample_rate)

Hopefully the image below clarifies how things are stored.  

wcl_def

### CODE REVIEW

**Comment by CaptKrasno** [comment]
@valschmidt @lauralindzey do you guys have any feedback on this?   I would like to get it changed or merged as necessary ASAP and it is required for upcoming projects.

File: acoustic_msgs/msg/MultibeamWatercolumn.msg [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+# May be empty if not reported
+# reported in radians
+float32[] tx_beamwidth
+float32[] rx_beamwidth
+
+bool      is_bigendian
+
+uint32  DTYPE_UINT8 = 0
+uint32  DTYPE_INT8 = 1
+uint32  DTYPE_INT16 = 2
```
Comment (lindzey) [inline acoustic_msgs/msg/MultibeamWatercolumn.msg:49]: UINT16?

File: acoustic_msgs/msg/MultibeamWatercolumn.msg [hunk 2]
```diff
@@ -0,0 +1,60 @@
+
+# The header time should be the time the ping was transmitted, not
+# the time the message was received.
+# Note that these two could potentially be several seconds apart,
+# especially in deep water.
+std_msgs/Header header
+
+float32 sound_speed   # m/s
```
Comment (rolker) [inline acoustic_msgs/msg/MultibeamWatercolumn.msg:8]: Add comment that this is sound speed typically measured at the head, not to be confused with an average watercolumn sound speed for example.

**Comment by rolker** [comment]
Seems like the only ways I can see to determine beam count are to look at the size of rx_angles or the size of image.data and account for data type and sampes_per_beam.

Should we add a beam_count member or at document the prefered method of determining the number of beams?

**Comment by CaptKrasno** [comment]
I think we had one in there initially and decided to take it out during our phone conversation.  I wouldn't be opposed to adding it back, it is convenient.   Maybe add it to the SonarImageData sub message?  That way a SonarImageData could be parsed without the wrapper Datatype.

-Kris

**Comment by rolker** [comment]
Yeah, adding it to SonarImageData does make sense. As is, it can't be processed independently.

**Comment by lauralindzey** [comment]
@CaptKrasno -- any chance that you could make the minor revisions that are still outstanding on this PR, and then we can merge it?

I'm currently installing your Norbit driver on NUI, and am eager to use these messages for our cruise =)

File: acoustic_msgs/msg/PingInfo.msg [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+float32 frequency
+
+# Speed of sound (m/s) used to calculate ranges;
+# Set to 0 if unavailable
+float32 sound_speed
+
+# Sonar reported -3db beamwidths
+# May be empty if not reported
+# reported in radians
+float32[] tx_beamwidths
```
Comment (lauralindzey) [inline acoustic_msgs/msg/PingInfo.msg:12]: @CaptKrasno -- are you willing to add a note describing these in a bit more detail? e.g. a roboticist wanting to get started using these may not know how transmit/receive beamforming is performed, and thus won't know which axes these correspond to.

File: acoustic_msgs/msg/MultibeamWatercolumn.msg (hunk 1 above)
Comment (lauralindzey) [reply acoustic_msgs/msg/MultibeamWatercolumn.msg]: Fixed when this was pulled into SonarImageData.

File: acoustic_msgs/msg/SonarRanges.msg [hunk 4]
```diff
@@ -15,31 +15,7 @@ std_msgs/Header header
 # than one thing to flag beams as bad. To check if a beam is good, simply
```
Comment (lauralindzey) [inline acoustic_msgs/msg/SonarRanges.msg:15]: (comment for line 5; it wouldn't let me select it ...)

Do you want to add attributions to the other institutions who have worked on this, or just delete this line?
(I'm ambivalent, but as it stands now, it's inaccurate)

File: acoustic_msgs/msg/SonarRanges.msg [hunk 5]
```diff
@@ -15,31 +15,7 @@ std_msgs/Header header
...
-# the number of beams for this sonar.
-#
-# Some fields are optional.  Those should have 0 length.
-
-# Flag values for every beam
-uint8[] flag
-
-# Truly raw multibeam data uses travel times rather than ranges.
-float32[] two_way_travel_time
+DetectionFlag[] flags
```
Comment (lauralindzey) [inline acoustic_msgs/msg/SonarRanges.msg:18]: Again, this comment is for a different line that Github wouldn't let me select. GRRR.

Did you intentionally leave PingInfo out of this message, or was that just an oversight?

**Comment by lauralindzey** [comment]
Also, I just want to say, that after not thinking about these messages for a while and then coming back, I **REALLY** like how you've laid them out. It makes sense to me, and works for both of the multibeam projects I have right now. (Imaging sonar reconstruction with an Oculus, and mapping with a Norbit.)

**Comment by rolker** [comment]
I second Laura's friendly nudge to have this merged in before this cruise gets underway. I'll be using the format to collect EK80 data from Drix and hopefully from Drix's other sonars as well, so now's a great time to freeze this message format.

**Comment by lauralindzey** [comment]
Also, just as a FYI, I'm working on a PR to your norbit driver using these message formats.
I'm trying to get it all done tonight so we can test it during tomorrow's NUI dunk test. 

(Yes, I'm kicking myself for not pushing to get this merged earlier and giving you the chance to do that work, but oh well ...)

**Comment by rolker** [comment]
Another clarification we could make in the comments is about the tx_delays and tx_angles arrays. Can they be empty if all values are zero? Even the rx_angles could potentially be empty in the case of a single beam sonar :)

**Comment by CaptKrasno** [comment]
@lindzey I have added the number of beams field to the ping info message.   I will do some documentation soon.

**Comment by CaptKrasno** [comment]
@lauralindzey @rolker 

I've sent you an email discussing some issues with coordinate convention and catching you up on the driver.    Let me know your thoughts ASAP and we can get this resolved.

**Comment by lauralindzey** [comment]
Merging!

We know of some documentation improvements that we should make, but I'm going to go ahead and merge this so we can start using it. The documentation can come in a separate PR.

**Pull request merged**